src/test/java/com/pylon/spokestack/google/GoogleSpeechRecognizerTest.java

```diff
@@ -0,0 +1,152 @@
+        );
+        client.getResponses().onCompleted();
+        assertEquals("test", context.getTranscript());
+        assertEquals(0.75, context.getConfidence());
+        assertEquals(SpeechContext.Event.RECOGNIZE, this.event);
+
+        // shutdown
+        recognizer.close();
+        verify(client.getStub()).close();
+    }
+
+    @Test
+    @SuppressWarnings("unchecked")
+    public void testError() throws Exception {
+        SpeechContext context = createContext();
+        MockSpeechClient client = spy(MockSpeechClient.class);
+        GoogleSpeechRecognizer recognizer =
+            new GoogleSpeechRecognizer(createConfig(), client);
+
+        // trigger recognition
+        context.setActive(true);
+        recognizer.process(context, context.getBuffer().getLast());
+        context.setActive(false);
+        recognizer.process(context, context.getBuffer().getLast());
+
+        // inject fault
+        client.getResponses().onError(new Exception("test error"));
+        assertEquals("test error", context.getError().getMessage());
+        assertEquals("", context.getTranscript());
+        assertEquals(0, context.getConfidence());
+        assertEquals(SpeechContext.Event.ERROR, this.event);
+    }
+
+    private SpeechConfig createConfig() {
+        SpeechConfig config = new SpeechConfig();
+        config.put("google-credentials", "{}");
+        config.put("sample-rate", 16000);
+        config.put("locale", "en-US");
+        return config;
+    }
+
+    private SpeechContext createContext() {
+        SpeechContext context = new SpeechContext();
+        context.addOnSpeechEventListener(this);
+
+        context.attachBuffer(new LinkedList<ByteBuffer>());
+        for (int i = 0; i < 3; i++)
+            context.getBuffer().addLast(ByteBuffer.allocateDirect(320));
+
+        return context;
+    }
+
+    public void onEvent(SpeechContext.Event event, SpeechContext context) {
+        this.event = event;
+    }
+
+    @SuppressWarnings("unchecked")
+    private static class MockSpeechClient extends SpeechClient {
+        private final BidiStreamingCallable callable;
+        private final ApiStreamObserver requests;
+
+        public MockSpeechClient() {
+            super(mock(SpeechStub.class));
+            this.callable = mock(BidiStreamingCallable.class);
+            this.requests = mock(ApiStreamObserver.class);
+            when(getStub().streamingRecognizeCallable())
+                .thenReturn(this.callable);
+            when(this.callable.bidiStreamingCall(any(ApiStreamObserver.class)))
+                .thenReturn(this.requests);
+        }
+
+        public ApiStreamObserver getRequests() {
+            return this.requests;
+        }
+
+        public ApiStreamObserver getResponses() {
+            ArgumentCaptor<ApiStreamObserver> captor =
+                ArgumentCaptor.forClass(ApiStreamObserver.class);
+            verify(this.callable)
+                .bidiStreamingCall(captor.capture());
```

Comment:
actual lol

Reply:
I'm not sure who is the captor and who is the captee at this point.